src/libstd/error.rs

```diff
@@ -71,6 +74,14 @@ pub trait Error: Debug + Display {
     /// The lower-level cause of this error, if any.
     #[stable(feature = "rust1", since = "1.0.0")]
     fn cause(&self) -> Option<&Error> { None }
+
+    /// Get the `TypeId` of `self`
+    #[doc(hidden)]
+    #[unstable(feature = "core",
+               reason = "unclear whether to commit to this public implementation detail")]
+    fn type_id(&self) -> TypeId where Self: 'static {
```

Comment:
Thought about this some more, and we can get away without stabilizing `Reflect` with just a bit less flexibility - instead of implementing Error for `T: Reflect` in e.g. TryLockError, just implement for `T: Any`. It's slightly less flexible, but it's backwards compatible to generalize the impl when/if we do stabilize `Reflect` directly.